h/streamer/messages.py

```diff
@@ -129,35 +137,37 @@ def _generate_annotation_event(
     if message["src_client_id"] == socket.client_id:
         return None
 
-    # We don't send anything until we have received a filter from the client
-    if socket.filter is None:
-        return None
-
     # Don't sent annotations from NIPSA'd users to anyone other than that
     # user.
     if user_nipsad and socket.authenticated_userid != annotation.userid:
         return None
 
-    notification = {"type": "annotation-notification", "options": {"action": action}}
+    # The `prepare` function sets the active registry which is an implicit
+    # dependency of some of the authorization logic used to look up annotation
+    # and group permissions.
+    with pyramid.scripting.prepare(registry=socket.registry):
```

Comment:
[Docs for `prepare`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/scripting.html#pyramid.scripting.prepare):

> This function pushes data onto the Pyramid threadlocal stack (request and registry), making those objects 'current'. It returns a dictionary useful for bootstrapping a Pyramid application in a scripting environment.

I'm not sure the `prepare` is really intended for this use case, but it solves the problem of setting the current registry which various authorization checks implicitly depend upon, in particular calls to `principals_allowed_by_permission`.

Reply:
Pyramid. Black magic. Wave wand. Sure!